# [Цели исследования](#stage_0_1)

# [Описание данных](#stage_0_2)

# [1. Изучение таблиц](#stage_1)

# [2. Предобработка данных](#stage_2)

# [3. Оценка корректности проведения теста](#stage_3)

# [4. Исследователький анализ данных](#stage_4)

## [Промежуточный вывод](#stage_4_1)

# [5.Оценка результатов A/B-тестирования ](#stage_5)
## [Промежуточный вывод](#stage_5_1)

# [6. Выводы](#stage_6)

<a name="stage_0_1"></a>

## Цели исследования 

Проанализировать базу данных приобретенного сервиса для чтения книг по подписке и сформулировать ценностное предложение для нового продукта.

<a name="stage_0_2"></a>
## Описание данных

Таблица `books`  
Содержит данные о книгах:  
- `book_id` — идентификатор книги;  
- `author_id` — идентификатор автора;  
- `title` — название книги;  
- `num_pages` — количество страниц;  
- `publication_date` — дата публикации книги;  
- `publisher_id` — идентификатор издателя.  

Таблица `authors`  
Содержит данные об авторах:  
- `author_id` — идентификатор автора;  
- `author` — имя автора.  

Таблица `publishers`  
Содержит данные об издательствах:  
- `publisher_id` — идентификатор издательства;  
- `publisher` — название издательства;  

Таблица `ratings`  
Содержит данные о пользовательских оценках книг:  
- `rating_id` — идентификатор оценки;  
- `book_id` — идентификатор книги;  
- `username` — имя пользователя, оставившего оценку;  
- `rating` — оценка книги.  

Таблица `reviews`  
Содержит данные о пользовательских обзорах:  
- `review_id` — идентификатор обзора;  
- `book_id` — идентификатор книги;  
- `username` — имя автора обзора;  
- `text` — текст обзора.  

In [1]:
#Импортируем нужные библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [2]:
# функция для исследования таблиц
def table_analyze(table):
    display(pd.io.sql.read_sql(sql=text('''SELECT * FROM {} LIMIT 5'''.format(table)), con = engine.connect()))
    pd.io.sql.read_sql(sql=text('''SELECT * FROM {}'''.format(table)), con = engine.connect()).info()
    display(pd.io.sql.read_sql(sql=text('''SELECT * FROM {}'''.format(table)), con = engine.connect()).describe())
    print('*' *30)

<a name="stage_0_1"></a>

# 1. Изучение таблиц

Применяя функцию table_analyze рассмотрим общую информацию по таблицам.

In [3]:
name_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for name in name_list:
    print(f'Вывод общей информации по таблице `{name}`:')
    table_analyze(name)

Вывод общей информации по таблице `books`:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

******************************
Вывод общей информации по таблице `authors`:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

******************************
Вывод общей информации по таблице `publishers`:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

******************************
Вывод общей информации по таблице `ratings`:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

******************************
Вывод общей информации по таблице `reviews`:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

******************************


Импортировали необходимыке библиотеки и параметры доступа к базе данных, изучили общую информацию о данных.   

Видим, что пропуски отсутствуют. Типы данных также соответствуют значениям.

<a name="stage_2"></a>
# 2. Решение задания

1. Посчитаем, сколько книг вышло после 1 января 2000 года;

In [4]:
query = ''' 
            SELECT
                 COUNT(title) as books_count 
            FROM books AS b
            WHERE publication_date >=DATE('2000-01-01')
        '''
pd.io.sql.read_sql(sql=text(query), con = engine.connect())

books_count
0          821

После 1 января 2000 года включительно вышла 821 книга.

2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [5]:
query = ''' 
            WITH temp1 as (
            SELECT
                 book_id,
                 COUNT(review_id) as count_reviews
            FROM 
                reviews as r
            GROUP BY book_id
            ),
            temp2 as (
            SELECT
                 book_id,
                 AVG(rating) as avg_rating
            FROM 
                ratings as ra
            GROUP BY book_id
            )
            SELECT
                 b.book_id,
                 b.title,
                 sum(count_reviews) as count_reviews,
                 avg(avg_rating) as avg_rating
            FROM 
                books as b
            LEFT JOIN 
                    temp1 ON b.book_id = temp1.book_id
            LEFT JOIN 
                    temp2 ON b.book_id = temp2.book_id
            GROUP BY 1
            ORDER BY avg_rating DESC
        '''
pd.io.sql.read_sql(sql=text(query), con = engine.connect())

book_id                                              title  \
0        297                                         Hard Times   
1        902  The War of Art: Break Through the Blocks & Win...   
2        599  Stone of Farewell (Memory  Sorrow  and Thorn  #2)   
3         20              A Fistful of Charms (The Hollows  #4)   
4        625  The Adventures of Tom Sawyer and Adventures of...   
..       ...                                                ...   
995      915  The World Is Flat: A Brief History of the Twen...   
996      202                                      Drowning Ruth   
997      316                  His Excellency: George Washington   
998      371                                              Junky   
999      303                               Harvesting the Heart   

     count_reviews  avg_rating  
0              2.0        5.00  
1              2.0        5.00  
2              1.0        5.00  
3              2.0        5.00  
4              1.0        5.00  
..             ...         ...  
995            3.0        2.25  
996            3.0        2.00  
997            2.0        2.00  
998            2.0        2.00  
999            2.0        1.50  

[1000 rows x 4 columns]

Вывод отвечает на вопрос поставленной задачи

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [6]:
query = ''' 
            SELECT
                 p.publisher,
                 COUNT(book_id) as book_count
            FROM 
                publishers AS p
            JOIN 
                books as b ON p.publisher_id=b.publisher_id
            WHERE 
                num_pages > 50
            GROUP BY 1
            ORDER BY 2 DESC
        '''
pd.io.sql.read_sql(sql=text(query), con = engine.connect()).head()

publisher  book_count
0             Penguin Books          42
1                   Vintage          31
2  Grand Central Publishing          25
3          Penguin Classics          24
4          Ballantine Books          19

Наибольшее число книг (42 шт) выпустило издательство Penguin Books	

4. Определение автора с самой высокой средней оценкой книг;

Учитываем только книги с 50 и более оценками.

In [7]:
query = ''' WITH temp1 AS (SELECT 
                                book_id
                           FROM 
                                ratings
                           GROUP BY 
                                   book_id
                           HAVING 
                                COUNT(rating_id) >= 50),

                temp2 AS (SELECT 
                                b.author_id as a_id,
                                ROUND(AVG(r.rating),2) AS avg_rating
                       FROM 
                           temp1
                       JOIN 
                           books b ON b.book_id = temp1.book_id
                       JOIN 
                           ratings r ON r.book_id=temp1.book_id
                       GROUP BY 
                               b.author_id
                       ORDER BY 
                               avg_rating DESC)

          SELECT 
                temp2.a_id,
                a.author,
                temp2.avg_rating
          FROM 
                temp2
          JOIN
                authors a ON temp2.a_id=a.author_id
          ORDER BY avg_rating DESC
            
        '''
pd.io.sql.read_sql(sql=text(query), con = engine.connect())

a_id                                             author  avg_rating
0    236                         J.K. Rowling/Mary GrandPré        4.29
1    402                  Markus Zusak/Cao Xuân Việt Khương        4.26
2    240                                     J.R.R. Tolkien        4.25
3    376                                  Louisa May Alcott        4.19
4    498                                       Rick Riordan        4.08
5    621                                    William Golding        3.90
6    235                                      J.D. Salinger        3.83
7    630  William Shakespeare/Paul Werstine/Barbara A. M...        3.79
8    469           Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.79
9    372                                         Lois Lowry        3.75
10   106                                          Dan Brown        3.74
11   195            George Orwell/Boris Grabnar/Peter Škerl        3.73
12   554                                    Stephenie Meyer        3.66
13   311                                     John Steinbeck        3.62

Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré.

5. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [8]:
query = ''' 
            SELECT
                 ROUND(AVG(review_ids)) AS avg_reviews
            FROM 
                (SELECT
                        COUNT(review_id) AS review_ids
                        FROM
                            (SELECT
                                    username
                            FROM
                                ratings
                            GROUP BY 1
                            HAVING
                                COUNT(rating_id) > 48) as tmp1
            LEFT JOIN 
                    reviews as r ON tmp1.username = r.username
            GROUP BY 
                    tmp1.username)as tmp2
        '''
pd.io.sql.read_sql(sql=text(query), con = engine.connect())

avg_reviews
0         24.0

Средним количеством обзоров от пользователей, которые поставили больше 48 оценок, является 24

<a name="stage_5"></a>

# 5. Выводы

По результатам выполнения sql запросов мы имеем следующие данные:
1. После 1 января 2000 года включительно вышла 821 книга;    
2. Для каждой книги посчитано количество обзоров и средняя оценка;  
3. Наибольшее число книг (42 шт) выпустило издательство Penguin Books;  
4. Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré.  
5. Средним количеством обзоров от пользователей, которые поставили больше 48 оценок, является 24